core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/Region.scala

```diff
@@ -1,36 +1,71 @@
 package edu.uci.ics.amber.engine.architecture.scheduling
 
+import edu.uci.ics.amber.engine.architecture.deploysemantics.PhysicalOp
 import edu.uci.ics.amber.engine.architecture.scheduling.config.ResourceConfig
 import edu.uci.ics.amber.engine.common.virtualidentity.PhysicalOpIdentity
 import edu.uci.ics.amber.engine.common.workflow.PhysicalLink
+import org.jgrapht.graph.{DefaultEdge, DirectedAcyclicGraph}
+import org.jgrapht.traverse.TopologicalOrderIterator
+
+import scala.jdk.CollectionConverters.asScalaIteratorConverter
 
 case class RegionLink(fromRegion: Region, toRegion: Region)
 
-case class RegionIdentity(id: String)
+case class RegionIdentity(id: Long)
 
 case class Region(
     id: RegionIdentity,
-    physicalOpIds: Set[PhysicalOpIdentity],
+    physicalOps: Set[PhysicalOp],
     physicalLinks: Set[PhysicalLink],
     resourceConfig: Option[ResourceConfig] = None,
-    // operators whose all inputs are from upstream region.
-    sourcePhysicalOpIds: Set[PhysicalOpIdentity] = Set.empty,
-    // links to downstream regions, where this region generates blocking output.
+    // operators and links in part of the downstream region.
+    downstreamOps: Set[PhysicalOp] = Set.empty,
     downstreamLinks: Set[PhysicalLink] = Set.empty
 ) {
 
+  private val operators: Map[PhysicalOpIdentity, PhysicalOp] =
+    getEffectiveOperators.map(op => op.id -> op).toMap
+
+  @transient lazy val dag: DirectedAcyclicGraph[PhysicalOpIdentity, DefaultEdge] = {
+    val jgraphtDag = new DirectedAcyclicGraph[PhysicalOpIdentity, DefaultEdge](classOf[DefaultEdge])
+    (physicalOps ++ downstreamOps).foreach(op => jgraphtDag.addVertex(op.id))
+    (physicalLinks ++ downstreamLinks)
+      .filter(link => physicalOps.map(_.id).contains(link.fromOpId))
+      .foreach(link => jgraphtDag.addEdge(link.fromOpId, link.toOpId))
+    jgraphtDag
+  }
+  def topologicalIterator(): Iterator[PhysicalOpIdentity] = {
+    new TopologicalOrderIterator(dag).asScala
+  }
+
   /**
-    * Return all PhysicalOpIds that this region may affect.
+    * Return all PhysicalOps that this region may affect.
     * This includes:
     *   1) operators in this region;
-    *   2) operators not in this region but blocked by this region (connected by the downstream links).
+    *   2) operators not in this region but receiving input from by this region.
     */
-  def getEffectiveOperators: Set[PhysicalOpIdentity] = {
-    physicalOpIds ++ downstreamLinks.map(link => link.toOpId)
-  }
+  def getEffectiveOperators: Set[PhysicalOp] = physicalOps ++ downstreamOps
 
   def getEffectiveLinks: Set[PhysicalLink] = {
     physicalLinks ++ downstreamLinks
   }
 
+  def getEffectiveOperator(physicalOpId: PhysicalOpIdentity): PhysicalOp = {
+    operators(physicalOpId)
+  }
+
+  /**
+    * Effective source operators in a region are effective operators that have 0 effective input links.
+    */
+  def getEffectiveSourceOpIds: Set[PhysicalOpIdentity] = {
```

Comment:
I found the definition of "effective" is vague. In the region, we have operators and some extra weak-ref operators. It seems "effective" means "all". If that is the case, can we use "all" instead? Also, I think the source operators never include weak-ref operators. Is that true?